src/NimBLEClient.cpp

```diff
@@ -313,16 +314,42 @@ bool NimBLEClient::connect(const NimBLEAddress& address, bool deleteAttributes,
  * @return True on success.
  * @details This is a blocking function and should not be used in a callback.
  */
-bool NimBLEClient::secureConnection() const {
+bool NimBLEClient::secureConnection(bool async) {
     NIMBLE_LOGD(LOG_TAG, ">> secureConnection()");
 
+    if (!NimBLEDevice::m_synced) {
+        NIMBLE_LOGE(LOG_TAG, "Host reset, wait for sync.");
+        return false;
+    }
+
+    if (NimBLEDevice::isSecureInProgress()) {
+        NIMBLE_LOGE(LOG_TAG, "Secure already in progress");
+        return false;
+    }
+
+    m_asyncSecure = async;
+
+    // Set the secure in progress flag to prevent a scan from starting while securing.
+    NimBLEDevice::setSecureInProgress(true);
```

Comment:
I tried to copy the code structure from async connect since I thought this needed similar handling, but have trimmed out the unneeded checks.

Reply:
All good, scanning only needs to be prevented when a connection attempt is in progress as the stack needs to take over the receiver so it can respond to an advertisement with a connect operation, after that the communications are scheduled. Please test it shough if you can.

Reply:
Here's what I'm seeing so far, 2 peripherals which the esp32-s3 will connect to. Scanning while securing appears to be fine, however it looks like it's possible to try initiating a second secure call while the first has not yet completed. My application still has some remnants from when it was servicing one device until disconnect, so I'll keep testing over the weekend. I kept the log level on warning so it's easier to parse, let me know if you want a debug log.

```
I [ 77740][main.cpp:214] BLE adv (-90 db, phy 1 0) Name: , Address: f1:01:06:9c:ce:ea
I [ 77761][main.cpp:230] Create client, attempting to connect...
I [ 77762][main.cpp:260] Scan ended...
I [ 80013][main.cpp:145] Connected to f101069cceea, RSSI 0 (bond 0)
I [ 80015][main.cpp:481] Scan restart...
D [ 80236][main.cpp:207] Client requesting passkey...
D [ 80237][main.cpp:53] Generated sensor key 736384
I [ 80327][main.cpp:214] BLE adv (-93 db, phy 1 0) Name: SHIELD, Address: 68:6d:3a:bc:7c:65
I [ 80432][main.cpp:214] BLE adv (-41 db, phy 1 0) Name: , Address: e3:e4:5b:75:ad:36
I [ 80453][main.cpp:230] Create client, attempting to connect...
I [ 80454][main.cpp:260] Scan ended...
I [ 81736][main.cpp:145] Connected to e3e45b75ad36, RSSI 0 (bond 0)
I [ 81737][main.cpp:481] Scan restart...
E (81737) NimBLEClient: Update params error: 530, 
E (81748) NimBLEDevice: ble_gap_security_initiate: rc=6 
E (81748) NimBLEClient: secureConnection: failed to start rc=6
E [ 81759][main.cpp:157] Failed to secure, delete client
I [ 83267][main.cpp:260] Scan ended...
I [ 83269][main.cpp:481] Scan restart...
D [ 83284][main.cpp:188] Auth event 1 1 1 16
D [ 83284][main.cpp:192] Secured connection
```

Reply:
interesting, not sure why that return code is given... it should have bee EALREADY (2) if it was already started.